write-fonts/src/tables/glyf.rs

```diff
@@ -86,29 +88,291 @@ pub enum BadKurbo {
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Add a line segment to all contours
+    fn line_to(&mut self, pts: Vec<kurbo::Point>) {
+        assert_eq!(pts.len(), self.len());
+        for (i, pt) in pts.into_iter().enumerate() {
+            self.0[i].push(ContourPoint::on_curve(pt));
+        }
+    }
+
+    /// Add a quadratic curve segment to all contours
+    fn quad_to(&mut self, p0s: Vec<kurbo::Point>, p1s: Vec<kurbo::Point>) {
+        assert_eq!(p0s.len(), self.len());
+        assert_eq!(p1s.len(), self.len());
+        for (i, (p0, p1)) in p0s.into_iter().zip(p1s.into_iter()).enumerate() {
+            self.0[i].push(ContourPoint::off_curve(p0));
+            self.0[i].push(ContourPoint::on_curve(p1));
+        }
+    }
+
+    /// The total number of points in each interpolatable contour
+    fn num_points(&self) -> usize {
+        let n = self.0[0].len();
+        assert!(self.0.iter().all(|c| c.len() == n));
+        n
+    }
+
+    /// The first point in each contour
+    fn first(&self) -> Vec<&ContourPoint> {
+        self.0.iter().map(|v| v.first().unwrap()).collect()
+    }
+
+    /// The last point in each contour
+    fn last(&self) -> Vec<&ContourPoint> {
+        self.0.iter().map(|v| v.last().unwrap()).collect()
+    }
+
+    /// Remove the last point from each contour
+    fn remove_last(&mut self) {
+        self.0.iter_mut().for_each(|c| {
+            c.pop().unwrap();
+        });
+    }
+
+    /// Compute the intersection of all implied on-curve points and return their indices
+    fn impliable_oncurve_points(&self) -> HashSet<usize> {
+        let epsilon = f32::EPSILON as f64; // should we make it a parameter?
+        let mut result = HashSet::new();
+        if let Some((first, others)) = self.0.split_first() {
+            result = impliable_oncurve_points(first, epsilon);
+            for points in others {
+                result = result
+                    .intersection(&impliable_oncurve_points(points, epsilon))
+                    .copied()
+                    .collect();
+            }
+        }
+        result
+    }
+
+    /// Build the contours, dropping any on-curve points that can be implied in all contours
+    fn build(self) -> Vec<Contour> {
+        let drop = self.impliable_oncurve_points();
+        // Omit implied on-curve points when building the contours
+        let mut contours = Vec::with_capacity(self.len());
+        for points in self.0 {
+            contours.push(Contour(
+                points
+                    .into_iter()
+                    .enumerate()
+                    .filter_map(|(i, pt)| {
+                        if drop.contains(&i) {
+                            None
+                        } else {
+                            Some(CurvePoint::from(pt))
+                        }
+                    })
+                    .collect(),
+            ))
+        }
+        contours
+    }
```

Comment:
I find myself wondering if we can implement what the comment says more directly. Rather than allocating hash sets do a forward walk and filter if all points at index are implicit.

Reply:
#377 shows an exploratory implementation